### PR TITLE
Expandable recording rows for inline reports

### PR DESCRIPTION
**Pull request by andrewazores**
Rebased and reopened version of #73 after `react` branch was merged into `master`

### CODE REVIEW

**Comment by andrewazores** [comment]
@jiekang with the last two commits I've applied some pretty light and simple fixes to avoid re-renders of the reports. Clicking the action dropdown menu should no longer cause reports to re-render, and neither do timer updates if none of the recording values have changed. If any of the recording list data has changed (ie a new recording appears or disappears or changes running state), then currently all of the open reports do re-render. Likewise, opening or closing a report frame will cause all other open frames to also re-render. These cases are still because these actions cause state changes on the parent list component, so this causes the children components to be re-rendered. I'm looking into ways to avoid this re-render as well, but the fix might just be caching the reports.

**Comment by andrewazores** [comment]
@jiekang I just added a frontend caching layer and it still doesn't look great. The re-renders are still noticeable. I don't know how to work around it and not require a re-render at all though, at least not while using the PF DataList table component - even just checking the checkboxes counts as a state change on the parent table component, which then re-renders the table to reflect the checkbox states, including re-rendering any expanded reports. The report frames are already using the `useMemo` hook to hint that they only need to be re-rendered if some other direct pieces of relevant state change, but it isn't working how I had hoped.

So tl;dr I think getting the report frames to avoid re-renders the way we would ideally prefer is going to take a good bit of thought and likely mean rewriting our own table component or something with this particular use case in mind. I'm not sure how worth it that is right now. What do you think?

**Comment by andrewazores** [comment]
Removing the `iframe` and instead using a simple `<div dangerouslySetInnerHTML={innerHtml} />`, where the `innerHtml` is populated by the cache-enabled ReportService seems to completely fix the performance issues of a re-render for frontend cached reports (although re-renders still occur when the table state is modified). I'm experimenting with this approach, but it will take a bit of thinking and extra work because the `iframe` gave us some useful things for free - ex. allowing the report to execute their own independent inline script which does things like hide the OK results and generally update the appearance, and includes some CSS as well that would need to somehow be extract and re-applied to the stripped out report contents (ie after removing the scripts in the `<head>`), and also allows for nice automatic scroll handling. If this ends up seeming like a workable approach then I think it would make the most sense to perform these transformations on the backend, maybe with a query parameter like `?inline=true`. If `true` then the transformations are applied so that the report content can be simply displayed inline when the table is expanded. Otherwise no transformations are applied, and the report is served as-is, which would be used for when the user wishes to download/view the report in full rather than viewing it inline.

**Comment by andrewazores** [comment]
If the caching approach used for stopped and archived recordings works well enough, then I'll look into applying a similar caching technique for active recordings as well, and invalidating their cache entries after some relatively short period of time. This will be more as a debounce than a real caching or rate-limiting, which still needs to be implemented on the backend as well.

**Comment by jiekang** [comment]
I've been testing the changes out. It seems there are some potential issues with opening/closing the reports over a long period of time, especially with continuous recordings in the mix. Reproducing what I saw is a pain, but eventually one of the reports which had previously been showing fine was in loading state basically forever (over my lunch break).

At the least, it looks like reports for stopped recordings don't reload when opening or closing rows. Continuous ones still do though...

For the use of memo hook, maybe you can submit a custom prop comparison function to decide when to re-render?

Also I think the whole spinner plus report depending on load state should be separated into the ReportFrame with useEffect + useState over useMemo since we want the spinner while it's waiting for the report from the server. Maybe that will help the overall issue...

**Comment by jiekang** [comment]
Oh maybe it's an issue in the server? I just saw:

```
[ERROR] io.vertx.ext.web.handler.impl.HttpStatusException: Internal Server Error
Caused by: java.io.IOException: java.lang.IllegalArgumentException: Unknown stream identifier 21
	at org.openjdk.jmc.rjmx.services.jfr.internal.JfrRecordingInputStreamV2.invokeOperation(JfrRecordingInputStreamV2.java:206)
	at org.openjdk.jmc.rjmx.services.jfr.internal.JfrRecordingInputStreamV2.closeStream(JfrRecordingInputStreamV2.java:159)
	at org.openjdk.jmc.rjmx.services.jfr.internal.JfrRecordingInputStreamV2.close(JfrRecordingInputStreamV2.java:145)
	at com.redhat.rhjmc.containerjfr.net.web.handlers.TargetReportGetHandler.handleReportPageRequest(TargetReportGetHandler.java:122)
	at com.redhat.rhjmc.containerjfr.net.web.handlers.TargetReportGetHandler.handleAuthenticated(TargetReportGetHandler.java:107)
	at com.redhat.rhjmc.containerjfr.net.web.handlers.AbstractAuthenticatedRequestHandler.handle(AbstractAuthenticatedRequestHandler.java:74)
	at com.redhat.rhjmc.containerjfr.net.web.handlers.AbstractAuthenticatedRequestHandler.handle(AbstractAuthenticatedRequestHandler.java:53)
	at io.vertx.ext.web.impl.BlockingHandlerDecorator.lambda$handle$0(BlockingHandlerDecorator.java:48)
	at io.vertx.core.impl.ContextImpl.lambda$executeBlocking$2(ContextImpl.java:316)
	at io.vertx.core.impl.TaskQueue.run(TaskQueue.java:76)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: javax.management.RuntimeMBeanException: java.lang.IllegalArgumentException: Unknown stream identifier 21
	at java.management/com.sun.jmx.interceptor.DefaultMBeanServerInterceptor.rethrow(DefaultMBeanServerInterceptor.java:829)
	at java.management/com.sun.jmx.interceptor.DefaultMBeanServerInterceptor.rethrowMaybeMBeanException(DefaultMBeanServerInterceptor.java:842)
	at java.management/com.sun.jmx.interceptor.DefaultMBeanServerInterceptor.invoke(DefaultMBeanServerInterceptor.java:811)
	at java.management/com.sun.jmx.mbeanserver.JmxMBeanServer.invoke(JmxMBeanServer.java:801)
	at java.management.rmi/javax.management.remote.rmi.RMIConnectionImpl.doOperation(RMIConnectionImpl.java:1466)
	at java.management.rmi/javax.management.remote.rmi.RMIConnectionImpl$PrivilegedOperation.run(RMIConnectionImpl.java:1307)
	at java.management.rmi/javax.management.remote.rmi.RMIConnectionImpl.doPrivilegedOperation(RMIConnectionImpl.java:1399)
	at java.management.rmi/javax.management.remote.rmi.RMIConnectionImpl.invoke(RMIConnectionImpl.java:827)
	at java.base/jdk.internal.reflect.GeneratedMethodAccessor60.invoke(Unknown Source)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at java.rmi/sun.rmi.server.UnicastServerRef.dispatch(UnicastServerRef.java:359)
	at java.rmi/sun.rmi.transport.Transport$1.run(Transport.java:200)
	at java.rmi/sun.rmi.transport.Transport$1.run(Transport.java:197)
	at java.base/java.security.AccessController.doPrivileged(Native Method)
	at java.rmi/sun.rmi.transport.Transport.serviceCall(Transport.java:196)
	at java.rmi/sun.rmi.transport.tcp.TCPTransport.handleMessages(TCPTransport.java:562)
	at java.rmi/sun.rmi.transport.tcp.TCPTransport$ConnectionHandler.run0(TCPTransport.java:796)
	at java.rmi/sun.rmi.transport.tcp.TCPTransport$ConnectionHandler.lambda$run$0(TCPTransport.java:677)
	at java.base/java.security.AccessController.doPrivileged(Native Method)
	at java.rmi/sun.rmi.transport.tcp.TCPTransport$ConnectionHandler.run(TCPTransport.java:676)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:834)
	at java.rmi/sun.rmi.transport.StreamRemoteCall.exceptionReceivedFromServer(StreamRemoteCall.java:303)
	at java.rmi/sun.rmi.transport.StreamRemoteCall.executeCall(StreamRemoteCall.java:279)
	at java.rmi/sun.rmi.server.UnicastRef.invoke(UnicastRef.java:164)
	at jdk.remoteref/jdk.jmx.remote.internal.rmi.PRef.invoke(Unknown Source)
	at java.management.rmi/javax.management.remote.rmi.RMIConnectionImpl_Stub.invoke(Unknown Source)
	at java.management.rmi/javax.management.remote.rmi.RMIConnector$RemoteMBeanServerConnection.invoke(RMIConnector.java:1021)
	at org.openjdk.jmc.rjmx.internal.MCMBeanServerConnection.invoke(MCMBeanServerConnection.java:259)
	at org.openjdk.jmc.rjmx.ConnectionToolkit.invokeOperation(ConnectionToolkit.java:196)
	at org.openjdk.jmc.rjmx.services.jfr.internal.FlightRecorderCommunicationHelperV2.invokeJfrOperation(FlightRecorderCommunicationHelperV2.java:91)
	at org.openjdk.jmc.rjmx.services.jfr.internal.FlightRecorderCommunicationHelperV2.invokeOperation(FlightRecorderCommunicationHelperV2.java:82)
	at org.openjdk.jmc.rjmx.services.jfr.internal.JfrRecordingInputStreamV2.invokeOperation(JfrRecordingInputStreamV2.java:197)
	... 13 more
Caused by: java.lang.IllegalArgumentException: Unknown stream identifier 21
	at jdk.management.jfr/jdk.management.jfr.StreamManager.getStream(StreamManager.java:49)
	at jdk.management.jfr/jdk.management.jfr.FlightRecorderMXBeanImpl.closeStream(FlightRecorderMXBeanImpl.java:160)
	at java.base/jdk.internal.reflect.GeneratedMethodAccessor108.invoke(Unknown Source)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at sun.reflect.misc.Trampoline.invoke(MethodUtil.java:71)
	at jdk.internal.reflect.GeneratedMethodAccessor37.invoke(Unknown Source)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at java.base/sun.reflect.misc.MethodUtil.invoke(MethodUtil.java:260)
	at java.management/com.sun.jmx.mbeanserver.ConvertingMethod.invokeWithOpenReturn(ConvertingMethod.java:193)
	at java.management/com.sun.jmx.mbeanserver.ConvertingMethod.invokeWithOpenReturn(ConvertingMethod.java:175)
	at java.management/com.sun.jmx.mbeanserver.MXBeanIntrospector.invokeM2(MXBeanIntrospector.java:117)
	at java.management/com.sun.jmx.mbeanserver.MXBeanIntrospector.invokeM2(MXBeanIntrospector.java:54)
	at java.management/com.sun.jmx.mbeanserver.MBeanIntrospector.invokeM(MBeanIntrospector.java:237)
	at java.management/com.sun.jmx.mbeanserver.PerInterface.invoke(PerInterface.java:138)
	at java.management/com.sun.jmx.mbeanserver.MBeanSupport.invoke(MBeanSupport.java:252)
	at java.management/javax.management.StandardMBean.invoke(StandardMBean.java:405)
	at java.management/com.sun.jmx.interceptor.DefaultMBeanServerInterceptor.invoke(DefaultMBeanServerInterceptor.java:809)
	at java.management/com.sun.jmx.mbeanserver.JmxMBeanServer.invoke(JmxMBeanServer.java:801)
	at java.management.rmi/javax.management.remote.rmi.RMIConnectionImpl.doOperation(RMIConnectionImpl.java:1466)
	at java.management.rmi/javax.management.remote.rmi.RMIConnectionImpl$PrivilegedOperation.run(RMIConnectionImpl.java:1307)
	at java.management.rmi/javax.management.remote.rmi.RMIConnectionImpl.doPrivilegedOperation(RMIConnectionImpl.java:1399)
	at java.management.rmi/javax.management.remote.rmi.RMIConnectionImpl.invoke(RMIConnectionImpl.java:827)
	at java.base/jdk.internal.reflect.GeneratedMethodAccessor60.invoke(Unknown Source)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at java.rmi/sun.rmi.server.UnicastServerRef.dispatch(UnicastServerRef.java:359)
	at java.rmi/sun.rmi.transport.Transport$1.run(Transport.java:200)
	at java.rmi/sun.rmi.transport.Transport$1.run(Transport.java:197)
	at java.base/java.security.AccessController.doPrivileged(Native Method)
	at java.rmi/sun.rmi.transport.Transport.serviceCall(Transport.java:196)
	at java.rmi/sun.rmi.transport.tcp.TCPTransport.handleMessages(TCPTransport.java:562)
	at java.rmi/sun.rmi.transport.tcp.TCPTransport$ConnectionHandler.run0(TCPTransport.java:796)
	at java.rmi/sun.rmi.transport.tcp.TCPTransport$ConnectionHandler.lambda$run$0(TCPTransport.java:677)
	at java.base/java.security.AccessController.doPrivileged(Native Method)
	at java.rmi/sun.rmi.transport.tcp.TCPTransport$ConnectionHandler.run(TCPTransport.java:676)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	... 1 more
```

**Comment by andrewazores** [comment]
Yea, eventually when recordings get very large it starts to take a very long time for the backend to create the report. I'm not sure what caused that backend error you saw, it sounds like somehow the recording was lost in the target? I again think that this is a backend issue that is being exposed through testing functionality in the frontend.

Reports for stopped recordings are still re-rendering, but they reports are cached within the client's memory as of a few commits back in this PR, so there is no backend report generation and no network round trip time and so it happens much more quickly. Continuous recordings aren't cached because the cache would be immediately invalid so they do still suffer from backend generation time and network round trip time.

I've made the change to move the spinner into the report frame component, which is a nice refactor but doesn't make any functional difference.

I don't think how the data list currently works is really compatible with the custom logic for deciding when to re-render a component. We're already using function components, which only trigger a re-render when their state or props change. The problem, as far as I can tell, is that the data list works by its own state (the DataListItem children) being mutated in order to add/remove rows on expand/collapse. Mutating the data list's state causes the data list to be re-rendered with a new list of children, which I think causes all of the children to be destroyed and recreated with their corresponding new props/state - not preserving the existing children and allowing them to flag whether or not they themselves need to be re-rendered.

**Comment by andrewazores** [comment]
Here's a radical idea - maybe we should only allow reports to be generated for archived recordings? Then the backend can trivially decide whether it's appropriate to cache reports (ie, it's always acceptable to cache reports for archived recordings), and it's also much easier to determine when cache entries should be deleted, since we're entirely in control of both the archived recordings and their cached reports. The frontend can still also apply its own in-memory caching layer as well to further reduce latency.

**Comment by jiekang** [comment]
Right, as far as toggling visible or invisible list content, it does make sense that when anything is toggled in the list, the whole list re-renders on screen.

It does not make sense that when the entire list is re-rendered, stopped active, or archived recordings which are currently opened have a request made to the backend to get the html contents again. The continuous active or running active recordings are another story. And the running recordings which are opened which then stop and then get opened again... I think we'll want to look at expanding on the UX specifically for these scenarios.

From the sounds of it, the request to backend is no longer made again for the stopped recordings which are already open, right?

**Comment by andrewazores** [comment]
Only one request is made for any stopped or archived recording. This response is then cached and used for any subsequent re-renders for that same report. This is the case regardless of why the request is made for the report, whether it's a different part of the list being modified or the specific row in question being expanded and collapsed. All stopped or archived recording reports can be cached, and all requests that can be served a cached result are.

There is even an open PR on the backend for archived reports so that those are cached to disk in the backend, so if frontend clients request a report for an archived report it is really only ever generated once.

**Comment by jiekang** [comment]
Okay, it wasn't clear to me from the comments as well as the PR continuing to be in draft. From my point of view, the PR was fine to go at that point then. Sorry I didn't look over it very closely until now. Did you have other changes you wanted to put in? Otherwise feel free to mark it as ready for review and I can approve it.
